rslib/src/import_export/gather.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use itertools::Itertools;
+
+use crate::{
+    decks::NormalDeck,
+    io::filename_is_safe,
+    latex::extract_latex,
+    prelude::*,
+    revlog::RevlogEntry,
+    text::{extract_media_refs, extract_underscored_css_imports, extract_underscored_references},
+};
+
+#[derive(Debug, Default)]
+pub(super) struct ExportData {
+    pub(super) decks: Vec<Deck>,
+    pub(super) notes: Vec<Note>,
+    pub(super) cards: Vec<Card>,
+    pub(super) notetypes: Vec<Notetype>,
+    pub(super) revlog: Vec<RevlogEntry>,
+    pub(super) decks_configs: Vec<DeckConfig>,
+    pub(super) media_paths: HashSet<PathBuf>,
+}
+
+impl ExportData {
+    pub(super) fn gather_data(
+        &mut self,
+        col: &mut Collection,
+        search: impl TryIntoSearch,
+        with_scheduling: bool,
+    ) -> Result<()> {
+        self.notes = col.gather_notes(search)?;
+        self.cards = col.gather_cards()?;
+        self.decks = col.gather_decks()?;
+        self.notetypes = col.gather_notetypes()?;
+
+        if with_scheduling {
+            self.revlog = col.gather_revlog()?;
+            self.decks_configs = col.gather_deck_configs(&self.decks)?;
+        } else {
+            self.remove_scheduling_information(col);
+        };
+
+        col.storage.clear_searched_notes_table()?;
+        col.storage.clear_searched_cards_table()
+    }
+
+    pub(super) fn gather_media_paths(&mut self) {
+        let mut inserter = |name: &str| {
+            if filename_is_safe(name) {
+                self.media_paths.insert(PathBuf::from(name));
+            }
+        };
+        let svg_getter = svg_getter(&self.notetypes);
+        for note in self.notes.iter() {
+            gather_media_paths_from_note(note, &mut inserter, &svg_getter);
+        }
+        for notetype in self.notetypes.iter() {
+            gather_media_paths_from_notetype(notetype, &mut inserter);
+        }
+    }
+
+    fn remove_scheduling_information(&mut self, col: &Collection) {
+        self.remove_system_tags();
+        self.reset_deck_config_ids();
+        self.reset_cards(col);
+    }
+
+    fn remove_system_tags(&mut self) {
+        // TODO: case folding? child tags?
+        for note in self.notes.iter_mut() {
+            note.tags = std::mem::take(&mut note.tags)
+                .into_iter()
+                .filter(|tag| !matches!(tag.as_str(), "marked" | "leech"))
+                .collect();
+        }
+    }
+
+    fn reset_deck_config_ids(&mut self) {
+        for deck in self.decks.iter_mut() {
+            if let Ok(normal_mut) = deck.normal_mut() {
+                normal_mut.config_id = 1;
+            } else {
+                // TODO: scheduling case
```

Comment:
I couldn't figure out how filtered decks are currently converted to regular ones if scheduling is included. 🤔 
Do I have to run the below code for all filtered decks in the scheduling case as well?

Reply:
Unintuitively, the current code handles this at import time (anki2.py:287, :356). It means renaming an apkg file to colpkg and importing it results in errors. Ideally we'd do it on the export side instead, but since we'll need to keep doing it on the import side to support legacy decks, we could get away with just doing that for now.

In the scheduling=reset case, we reset the config id to 1 so that the user's default deck preset/config is used. In the scheduling=keep case for normal cards, we preserve+copy the deck presets over, so the user can use this mode for moving/copying their cards/decks/presets from one profile to another. In the filtered deck case I guess the current code is discarding the deck config id? In the scheduling=keep case, it might be preferable if we preserved the filtered deck and made sure the home deck was included at the same time, as that would allow users to shuffle filtered decks between profiles as they can do with normal decks. However, the first paragraph may come back to bite us there - if we went down that path, we'd need to make sure we're not breaking legacy exports. Do you have any thoughts/ideas/preferences on what we do here? We could potentially defer this part until the code to deal with the importing side is implemented.

Reply:
Consider the case where a user has created two separate profiles, and then wants to follow the advice on https://faqs.ankiweb.net/synchronizing-multiple-profiles.html. They need some way of being able to transfer content from one profile to the other in order to be able to merge the profiles, and currently as filtered decks can't be transported, it means the user will need to recreate them and their settings in the importing profile. Not a big deal for the majority of users who either don't use filtered decks, or use them ephemerally. But some users have a bunch of filtered decks with complex queries, and ideally in that case we'd provide a way for them to transport them. But to be fair, it is a niche use case, and probably doesn't come up much.